crates/aptos-crypto/src/p256/p256_sigs.rs

```diff
@@ -0,0 +1,191 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! This file implements traits for ECDSA signatures over NIST-P256.
+
+use crate::{
+    p256::{P256PrivateKey, P256PublicKey, ORDER_HALF, ORDER_MINUS_ONE},
+    hash::CryptoHash,
+    traits::*,
+};
+use anyhow::{anyhow, Result};
+use aptos_crypto_derive::{DeserializeKey, SerializeKey};
+use core::convert::TryFrom;
+use serde::Serialize;
+use std::{cmp::Ordering, fmt};
+use signature::Verifier;
+use p256::NonZeroScalar;
+use p256::NistP256;
+use p256::elliptic_curve::Curve;
+
+use super::P256_SIGNATURE_LENGTH;
+
+/// A P256 signature
+#[derive(DeserializeKey, Clone, SerializeKey)]
+pub struct P256Signature(pub(crate) p256::ecdsa::Signature);
+
+impl private::Sealed for P256Signature {}
+
+impl P256Signature {
+    /// The length of the P256Signature
+    pub const LENGTH: usize = P256_SIGNATURE_LENGTH;
+
+    /// Serialize an P256Signature.
+    pub fn to_bytes(&self) -> [u8; P256_SIGNATURE_LENGTH] {
+        // The RustCrypto P256 `to_bytes` call here should never return a byte array of the wrong length
+        self.0.to_bytes().try_into().unwrap()
+    }
+
+    /// Deserialize an P256Signature, without checking for malleability
+    pub(crate) fn from_bytes_unchecked(
+        bytes: &[u8],
+    ) -> std::result::Result<P256Signature, CryptoMaterialError> {
+        match p256::ecdsa::Signature::try_from(bytes) {
+            Ok(p256_signature) => Ok(P256Signature(p256_signature)),
+            Err(_) => Err(CryptoMaterialError::DeserializationError),
+        }
+    }
+
+    /// return an all-zero signature (for test only)
+    #[cfg(any(test, feature = "fuzzing"))]
+    pub fn dummy_signature() -> Self {
+        Self::from_bytes_unchecked(&[0u8; Self::LENGTH]).unwrap()
+    }
+
+    /// Check for correct size and third-party based signature malleability issues.
+    /// This method is required to ensure that given a valid signature for some message under some
+    /// key, an attacker cannot produce another valid signature for the same message and key.
+    ///
+    /// We use the technique described in
+    /// [BIP146](https://github.com/bitcoin/bips/blob/master/bip-0146.mediawiki) to prevent
+    /// malleability of ECDSA signatures. Signatures comprise elements {R, S}, and S can be
+    /// enforced to be of canonical form by ensuring it is less than the order of the P256 curve
+    /// divided by 2. If this is not done, a value S > n/2 can be replaced by S' = n - S to form another distinct valid
+    /// signature, where n is the curve order. This check is not performed by the RustCrypto P256 library
```

Comment:
Here you contradict the comment you made in the tests, where you say it does check...

Reply:
Fixed, was computing the mauled non-canonical signature incorrectly